### ISSUE TITLE
Compiling `no_std` for `i686-pc-windows-gnu` ignores `panic=abort`

### ISSUE THREAD

**Issue by Irate-Walrus** [rust-lang/rust#133826]
When cross-compiling a `#![no_std]` crate for `i686-pc-windows-gnu` with `panic=abort` it includes unwinding symbols and its dependencies:

```text
Archive member included to satisfy reference by file (symbol)

/usr/lib/gcc/i686-w64-mingw32/13-win32/libgcc_eh.a(unwind-dw2-fde.o)
                              /home/irate-walrus/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-pc-windows-gnu/lib/rsbegin.o (__register_frame_info)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libmsvcrt.a(libmsvcrt_defs01135.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/libgcc_eh.a(unwind-dw2-fde.o) (abort)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libmsvcrt.a(libmsvcrt_defs01177.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/libgcc_eh.a(unwind-dw2-fde.o) (free)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libmsvcrt.a(libmsvcrt_defs01228.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/libgcc_eh.a(unwind-dw2-fde.o) (malloc)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libmsvcrt.a(libmsvcrt_defs01234.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/libgcc_eh.a(unwind-dw2-fde.o) (memcpy)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libmsvcrt.a(libmsvcrt_defs01279.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/libgcc_eh.a(unwind-dw2-fde.o) (strlen)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libmsvcrt.a(libmsvcrt_defh.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libmsvcrt.a(libmsvcrt_defs01135.o) (_head_lib32_libmsvcrt_def_a)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libmsvcrt.a(libmsvcrt_deft.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libmsvcrt.a(libmsvcrt_defh.o) (_lib32_libmsvcrt_def_a_iname)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libmingw32.a(lib32_libmingw32_a-tlsmcrt.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/libgcc_eh.a(unwind-dw2-fde.o) (_CRT_MT)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libkernel32.a(libkernel32s01416.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/libgcc_eh.a(unwind-dw2-fde.o) (_imp__Sleep@4)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libkernel32.a(libkernel32s00996.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/libgcc_eh.a(unwind-dw2-fde.o) (_imp__LeaveCriticalSection@4)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libkernel32.a(libkernel32s00898.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/libgcc_eh.a(unwind-dw2-fde.o) (_imp__InitializeCriticalSection@4)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libkernel32.a(libkernel32s00321.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/libgcc_eh.a(unwind-dw2-fde.o) (_imp__EnterCriticalSection@4)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libkernel32.a(libkernel32h.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libkernel32.a(libkernel32s01416.o) (_head_lib32_libkernel32_a)
/usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libkernel32.a(libkernel32t.o)
                              /usr/lib/gcc/i686-w64-mingw32/13-win32/../../../../i686-w64-mingw32/lib/../lib/libkernel32.a(libkernel32h.o) (_lib32_libkernel32_a_iname)
```

As `panic=abort` is specified in the crate's [Cargo.toml](https://github.com/Irate-Walrus/stardust-rs/blob/f85933c11ec40cb9bb78d2a0360fc31831b7a44d/Cargo.toml#L12C1-L12C16) I would expect that these dependencies would not be linked. This would also be consistent with the behavior when compiling for `x86_64-pc-windows-gnu` which does not link these symbols.

I have also attempted the compilation of the crate using the `Z build-std=core,alloc,panic_abort -Z build-std-features=panic_immediate_abort` flags which as not resolved this issue. I also define my own panic-handler and other required symbols in the [main.rs](https://github.com/Irate-Walrus/stardust-rs/blob/f85933c11ec40cb9bb78d2a0360fc31831b7a44d/stardust/src/main.rs#L22-L47)

The project can be found here: https://github.com/Irate-Walrus/stardust-rs .
### Meta

`rustc --version --verbose`:
```
rustc 1.84.0-nightly (b8c8287a2 2024-11-03)
binary: rustc
commit-hash: b8c8287a229cd79604aa84c25e1235fc78cd5f2e
commit-date: 2024-11-03
host: x86_64-unknown-linux-gnu
release: 1.84.0-nightly
LLVM version: 19.1.3
```

<details><summary>Backtrace</summary>
<p>

N/A, the project compiles but due to linkage of additional symbols crashes during runtime with an unrelated error.

</p>
</details>

**Comment by bjorn3** [comment]
The target spec for windows-gnu targets hard codes `-lgcc_eh`: https://github.com/rust-lang/rust/blob/2633e019298f8d9f671c8f3d0d3348ce6fa1b8aa/compiler/rustc_target/src/spec/base/windows_gnu.rs#L63

**Comment by Irate-Walrus** [comment]
I believe that these linker flags are passed for both the `x86_64-pc-windows-gnu` and `i686-pc-windows-gnu` targets? If so, why does this issue only appear for `i686`? In addition, despite these libraries being included, it is my understanding that they are discarded/not-linked if none of their symbols are being used.

**Comment by mati865** [comment]
Maybe it's somehow related to rtstartup files?
https://github.com/rust-lang/rust/tree/master/library/rtstartup

**Comment by Irate-Walrus** [comment]
There is a lot of symbol similarity there, and it would be included on only `i686-pc-windows-gnu`. In the project I have `-nostartfiles` [set](https://github.com/Irate-Walrus/stardust-rs/blob/f85933c11ec40cb9bb78d2a0360fc31831b7a44d/stardust/build.rs#L10). So my assumption would be that these shouldn't be included.

**Comment by bjorn3** [comment]
`-nostartfiles` stops gcc from passing those object files to the linker. However in the case of rustc, it is rustc itself which passes it's own version of those objects unconditionally. `-nostartfiles` doesn't have an effect on this.

I wonder if for your purposes compiling for x86_64-unknown-none and using `extern "win64" instead of extern "C" for calling C api's would work.

**Comment by Irate-Walrus** [comment]
> I wonder if for your purposes compiling for x86_64-unknown-none and using `extern "win64" instead of extern "C" for calling C api's would work.

I find this a little confusing as I am attempting to compile for `i686` rather than `x86_64` in this case. If I were to follow this advice (and it ran), I believe memory address calculations would be incorrect due to the use of `usize` and `extern "win64"` calls would fail as the windows APIs would be expecting `cdecl` or `stdcall` calling conventions.

**Comment by bjorn3** [comment]
> I find this a little confusing as I am attempting to compile for i686 rather than x86_64 in this case.

Right, forgot you were trying to compile for i686. Looks like we don't have a builtin -none target for i686. There is the option to define your own target if necessary however, but this does require nightly. The easiest way to get a working custom target is likely to run `rustc +nightly -Zunstable-options --print target-spec-js --target i686-pc-windows-gnu`, edit the output json as necessary, put it in a file and finally use `cargo build -Zbuild-std=core --target my_target.json`.

> Thanks for clearing this up. I assume this is the same behaviour for both x86_64-pc-windows-gnu and i686-pc-windows-gnu targets? In which case this is not causing the issue.

Yes

**Comment by mati865** [comment]
No, they are used only for i686.  You could try building for i686-pc-windows-gnullvm (you might need to download llvm-mingw). Windows-gnullvm targets differ in a few aspects from windows-gnu; one is not using rtstartup at all.

**Comment by Irate-Walrus** [comment]
>  The easiest way to get a working custom target is likely to run `rustc +nightly -Zunstable-options --print target-spec-js --target i686-pc-windows-gnu`, edit the output json as necessary, put it in a file and finally use `cargo build -Zbuild-std=core --target my_target.json`.

I've done this and removed the `pre-link-*` and `post-link-*` attributes resulting in a binary that both compiles and does not include the unwinding dependencies. This does introduce some changes to the `Makefile.toml` but that's perfectly okay. The compiled binary is still 2x times larger (which is odd, I assume something is forcing to include extra padding) but I'll have to investigate further time allowing.

**Comment by Irate-Walrus** [comment]
> The compiled binary is still 2x times larger (which is odd, I assume something is forcing to include extra padding) but I'll have to investigate further time allowing.

This is no longer an issue.  Thank you for you help and time spent. Do you think the option to disable these hard coded includes would be likely, or at that point it is a different target and as such no point?

**Comment by mati865** [comment]
If you can devise a solution that causes no problems, it would likely be accepted.  Since rtstartup is necessary only for i686-pc-windows-gnu (not for x86_64 or any *-windows-gnullvm), it will be difficult to find someone willing to spend time on this issue.

**Comment by Irate-Walrus** [comment]
All right, unfortunately that is above my capability at the moment, although hopefully that will change eventually. I've documented the solution we reached on [rust-lang](https://users.rust-lang.org/t/inclusion-of-lkernel32-and-others-when-compiling-no-std-for-i686-pc-windows-gnu/121551) and I'm happy for you to close this issue if you feel it is appropriate to do so e.g. if it is/is-not a bug.

**Comment by mati865** [comment]
Even if it will not be fixed, keeping this issue open might someday help someone else with a similar issue.
